hspec-core/src/Test/Hspec/Core/Spec/Vault.hs

```diff
@@ -0,0 +1,32 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Test.Hspec.Core.Spec.Vault (
+  Vault
+, setValue
+, getValue
+) where
+
+import           Prelude ()
+import           Test.Hspec.Core.Compat
+
+import           Data.Typeable
+
+newtype Vault = Vault [(TypeRep, Value)]
+  deriving (
+#if MIN_VERSION_base(4,11,0)
+    Semigroup,
+#endif
+    Monoid)
+
+data Value = forall v . Typeable v => Value v
+
+setValue :: Typeable v => v -> Vault -> Vault
+setValue v (Vault vs) = Vault $ (typeOf v, Value v) : vs
+
+getValue :: forall v. Typeable v => Vault -> Maybe v
+getValue (Vault vs) = lookup (typeOf (undefined :: v)) vs >>= fromValue
```

Comment:
Yes, good point, I [was using `typeRep` at first…](https://github.com/hspec/hspec/runs/8190254611?check_suite_focus=true#step:4:306)

I was about to add `typeRep` to our `.Compat` module yesterday, but then didn’t feel it’s worth the effort.

Reply:
Btw, with `typeRep` you can use `Maybe v` (the return value itself) instead of `Proxy v`, which is simplifying the type signature, not requiring `ScopedTypeVariables`, saving you an import and possibly even more clearly stating intention (“look up the value by return type”):

```haskell
getValue (Vault vs) = let r = lookup (typeRep r) vs >>= fromValue in r
```